### PR TITLE
Miscellaneous render speedups

### PR DESCRIPTION
**Pull request by mbargull**
### Description

Addresses parts of gh-5224 .

The proposed changes do not fix the actual underlying issues with redundancy in the recipe re-processing happening while rendering.
They do treat symptoms of it, though.
This is done by avoiding a couple of obvious redundant function calls, substituting `copy.deepcopy` by `pickle.dumps/loads`, and foremost memoizing the lower level `select_lines` and regular expression evaluations.

It is intentionally narrowly scoped and does not alter the overall rendering process to avoid potential breaking changes (i.e., this PR is meant to be able to be "merged whenever" with no further dependencies/conflicts).

For the `conda-forge/arrow-cpp-feedstock` case from gh-5224 it cut the the overall processing time to about a quarter.

### Checklist - did you ...

- [x] Add a file to the `news` directory ([using the template](../blob/main/news/TEMPLATE)) for the next release's release notes?
     
- ~~[ ] Add / update necessary tests?~~
- ~~[ ] Add / update outdated documentation?~~

### CODE REVIEW

File: conda_build/metadata.py [hunk 1]
```diff
@@ -109,6 +109,31 @@ def remove_constructor(cls, tag):
 # used to avoid recomputing/rescanning recipe contents for used variables
 used_vars_cache = {}
 
+# Placeholders singletons for os/os.environ used to memoize select_lines().
+_selector_placeholder_os = object()
+_selector_placeholder_os_environ = object()
+
+
+@lru_cache(maxsize=200)
```
Comment (mbargull) [inline conda_build/metadata.py:117]: These `maxsize=200` are more or less arbitrarily chosen.
The default is `128` IIRC, but I just wanted to specify some (reasonable) number such that the temptation to add `maxsize=None` later is lessened.
We may not want to use `maxsize=None` deliberately to avoid unbound resource usage downstream, e.,g. in case multiple recipes are rendered in the same process.

**Comment by beckermr** [comment]
Oh this is fabulous! Thank you!

File: conda_build/config.py [hunk 2]
```diff
@@ -822,9 +823,10 @@ def clean_pkgs(self):
 
     def copy(self):
         new = copy.copy(self)
-        new.variant = copy.deepcopy(self.variant)
+        # Use picke.loads(pickle.dumps(...) as a faster copy.deepcopy alternative.
+        new.variant = pickle.loads(pickle.dumps(self.variant, -1))
         if hasattr(self, "variants"):
-            new.variants = copy.deepcopy(self.variants)
+            new.variants = pickle.loads(pickle.dumps(self.variants, -1))
```
Comment (beckermr) [inline conda_build/config.py:829]: Do you want a try except here that reverts back to deepcopy n case something cannot be pickled?

Comment (mbargull) [reply conda_build/config.py:829]: Thanks for taking a look!
I don't think we should get anything in there that `pickle` won't be able to handle; but a try/except also wouldn't hurt, of course.

File: conda_build/config.py [hunk 3]
```diff
@@ -822,9 +823,16 @@ def clean_pkgs(self):
...
+        try:
+            new.variant = pickle.loads(pickle.dumps(self.variant, -1))
+        except:
+            new.variant = copy.deepcopy(self.variant)
         if hasattr(self, "variants"):
-            new.variants = copy.deepcopy(self.variants)
+            try:
+                new.variants = pickle.loads(pickle.dumps(self.variants, -1))
+            except:
+                new.variants = copy.deepcopy(self.variants)
```
Comment (jezdez) [inline conda_build/config.py:835]: Let's do this without adding pickle and a bare except

File: conda_build/metadata.py (hunk 1 above)
Comment (jezdez) [reply conda_build/metadata.py:117]: Memoizing compiled patterns might give a slight advantage here

File: conda_build/metadata.py [hunk 4]
```diff
@@ -109,6 +109,31 @@ def remove_constructor(cls, tag):
...
+    if isinstance(pattern, re.Pattern):
+        return pattern.match(text, flags)
+    return re.match(pattern, text, flags)
+
+
+@lru_cache(maxsize=200)
+def re_search(pattern, text, flags=0):
+    if isinstance(pattern, re.Pattern):
+        return pattern.search(text, flags)
+    return re.search(pattern, text, flags)
```
Comment (jezdez) [inline conda_build/metadata.py:135]: Not a fan of this repetition and the pattern of having to cache regex

File: conda_build/metadata.py [hunk 5]
```diff
@@ -268,6 +293,58 @@ def eval_selector(selector_string, namespace, variants_in_place):
...
+            if isinstance(v, (list, set, tuple)):
+                namespace_copy[k] = tuple(
+                    tuple(e) if isinstance(e, (list, set)) else e for e in v
+                )
+        namespace_tuple = tuple(namespace_copy.items())
+        # Raise TypeError if anything in namespace_tuple is not hashable.
+        hash(namespace_tuple)
+    except TypeError:
+        return _select_lines(data, namespace, variants_in_place)
+    return _select_lines_memoized(data, namespace_tuple, variants_in_place)
```
Comment (jezdez) [inline conda_build/metadata.py:328]: This needs tests to verify that it's doing the right thing

**Comment by mbargull** [comment]
> Another case of a hyper optimization by adding a bunch of hacks, notably adding pickle into the system where there is no need before. 

Please adjust you attitude here. I will not continue to work on this given such preface.

These optimizations are hacks around the actual problem, yes (see top comment).
The actual problem has not been addressed because of hard to reason about code flow with many places touching global or semi-global state; which is why this is scoped to functions that do not interact with these (semi-)global states.

> notably adding pickle into the system where there is no need before.

There would be no need for this if `.config.Config.copy` would not be called an excessive amount of time (same code flow/state problem as above). Unless the number of calls is reduced, one needs to replace `deepcopy` with something else like `pickle`. I'm not fan of using `pickle` here (still imposes high resource usage) but also not a fan of having `deepcopy` here at all; if you have better or at least comparable solutions to this, please do apply them.

Analyzing the root causes and untangling the code flow paths is a much more elaborate work that I am not willing to do (as it enters the realm of questionable effort given that resources could be directed to the newer recipe format instead).

----
These changes have not been made out of fun, but because it poses an non-negligible strain on resources (compute and maintainer "wait" times) for conda-forge (that I have not been aware of before), see the case laid out in https://github.com/conda/conda-build/issues/5224 .

If you do want alternative implementations to the proposed ones, please profile this yourself or get someone to do so (I've used `py-instrument` for sampling and `yappi` for tracing -- advice: only use `yappi` with changes from this PR or on a small test case than `conda-smithy rerender` of `conda-forge/arrow-cpp-feedstock`); it's otherwise hard to reason about the impact on such changes.

**Comment by beckermr** [comment]
Hey @jezdez! Sorry I have been MIA in the conda world for a while. 

It may be useful to detail the pain points on render times throughout conda-forge. This is important context behind this PR. @mbargull mentioned a few of the locations, but I can detail more.

 - As @mbargull mentioned, long rerenders are becoming more common as we expand our build matrices and for certain things like compilers. This stresses our automated rerendering service and causes usability issues.
 - The conda-forge autotick bot spends a significant amount of time rerendering feedstocks. The long rerenders rate limit and stress the bot's infrastructure.
 - Every conda-forge build of an actual package also renders the recipe. Summed over all of the packages we build, even small changes in performance will have a big effect on our CI usage.
 
So I strongly disagree with the characterization of this PR as hyper-optimization. I do understand that code like this brings in additional maintenance burdens. @mbargull has been a long-term contributor to conda-build and is active in the community. I think he'll be around to help on maintenance here, but he should speak for himself.

**Comment by dholth** [comment]
For reference, Python's built-in [re cache](https://github.com/python/cpython/blob/3.12/Lib/re/__init__.py#L271-L278) appears to have 512 and 256 entries.

**Comment by mbargull** [comment]
@beckermr, thanks for giving more details!
You are far more involved with this day-to-day workings at conda-forge and knowledgeable about what this impacts.
The motivation for this PR is mostly an empathy thing realizing that you and the others have to deal with it...
To me personally, this is actually very low priority since I'm at most tangential (if at all) affected by this.
Which is also to say: IMO, the decision on when to include these or similar changes (`24.5` already? this year? next year?) is somewhat dependent how the affected parties push for it.

> @mbargull has been a long-term contributor to conda-build and is active in the community. I think he'll be around to help on maintenance here, but he should speak for himself.

Yes, happy to do so -- but even happier to rip such workarounds out again whenever they cease to be relevant!

**Comment by jezdez** [comment]
> > Another case of a hyper optimization by adding a bunch of hacks, notably adding pickle into the system where there is no need before.
> 
> Please adjust you attitude here. I will not continue to work on this given such preface.

As context for everyone, while preparing conda 24.3.0 yesterday, I found https://github.com/conda/conda/pull/13568 having been merged with little to no discussion which introduced a custom version parser into the recently added conda deprecations system, diverging from a standard parser. This is what I referred to above as "another case" related to inconvenient tradeoffs.

Generally speaking, I'm trying to encourage to follow engineering practices that solve the core problems rather than pile hacks onto hacks. I'll try harder to elaborate why I think some tradeoffs are not worth making in that context.

> These optimizations are hacks around the actual problem, yes (see top comment). The actual problem has not been addressed because of hard to reason about code flow with many places touching global or semi-global state; which is why this is scoped to functions that do not interact with these (semi-)global states.
> 
> > notably adding pickle into the system where there is no need before.
> 
> There would be no need for this if `.config.Config.copy` would not be called an excessive amount of time (same code flow/state problem as above). Unless the number of calls is reduced, one needs to replace `deepcopy` with something else like `pickle`. I'm not fan of using `pickle` here (still imposes high resource usage) but also not a fan of having `deepcopy` here at all; if you have better or at least comparable solutions to this, please do apply them.

To be clear, giving feedback of requiring changes to this PR does not imply that I have to provide a better solution instead. I agree that it is worth documenting the performance bottleneck that you're seeing (which you started in #5224, thank you!), so we can make the appropriate fixes, without having to compromise on code quality. You know that's how we gotten into this mess in the first place.

> Analyzing the root causes and untangling the code flow paths is a much more elaborate work that I am not willing to do (as it enters the realm of questionable effort given that resources could be directed to the newer recipe format instead).

I totally agree! This is significant work and it's not helped by patching over it again.

> These changes have not been made out of fun, but because it poses an non-negligible strain on resources (compute and maintainer "wait" times) for conda-forge (that I have not been aware of before), see the case laid out in #5224 .

I understand the context as I've read the ticket before making the code review, and take my responsibilities seriously, just like your contribution. I'm sorry if my review didn't live up to that.

> If you do want alternative implementations to the proposed ones, please profile this yourself or get someone to do so (I've used `py-instrument` for sampling and `yappi` for tracing -- advice: only use `yappi` with changes from this PR or on a small test case than `conda-smithy rerender` of `conda-forge/arrow-cpp-feedstock`); it's otherwise hard to reason about the impact on such changes.

Thanks for the advise, I'll add this to #5224, if you don't mind since it would enable others to pick this up as well.

**Comment by jezdez** [comment]
> Hey @jezdez! Sorry I have been MIA in the conda world for a while.

Hello! No apologies needed, as always :)

> It may be useful to detail the pain points on render times throughout conda-forge. This is important context behind this PR. @mbargull mentioned a few of the locations, but I can detail more.
> 
> * As @mbargull mentioned, long rerenders are becoming more common as we expand our build matrices and for certain things like compilers. This stresses our automated rerendering service and causes usability issues.
> * The conda-forge autotick bot spends a significant amount of time rerendering feedstocks. The long rerenders rate limit and stress the bot's infrastructure.
> * Every conda-forge build of an actual package also renders the recipe. Summed over all of the packages we build, even small changes in performance will have a big effect on our CI usage.

This is good context, which was missing in #5224, thank's for elaborating here!

> So I strongly disagree with the characterization of this PR as hyper-optimization. I do understand that code like this brings in additional maintenance burdens. @mbargull has been a long-term contributor to conda-build and is active in the community. I think he'll be around to help on maintenance here, but he should speak for himself.

I have to disagree that it's been effective for conda and conda-build to move solving fundamental problems to a later time, it's made both increasingly hard to maintain, maybe in some areas even impossible as @mbargull alluded in relation to the flow of the application logic.

**Comment by dholth** [comment]
[The deepcopy code](https://github.com/python/cpython/blob/main/Lib/copy.py#L135-L158) is a slow way to copy Python objects, json loads/dumps or pickle are known optimizations. We used the [technique in conda-index](https://github.com/conda/conda-index/blob/42b55cba0b50fe3ad7f54543e6ba2e000971954b/conda_index/index/__init__.py#L346-L347) to save time. In an old stackoverflow post e.g. the pickle loop is 2.7x faster than copy.deepcopy.

**Comment by mbargull** [comment]
@jezdez, thanks for elaborating more.
To explain my averse reaction to your initial review: The main review comment read more as a blanket statement which to me gave the overall issue (and how we have to tackle it) less recognition than warranted and bode a likely less productive and strenuous discussion (which I'm not eager to invest my time in).

To be clear from my side: None of these memoization approaches (in this PR and already elsewhere in the code base) are what I *want* us to have.
The problem we are facing is that we *would want* actual fixes, but coming up with those requires an unknown amount of work (that I can't provide at the moment and I don't know who else would do/sponsor such work).
Since `conda-build` is continuously used, we'd have to decide whether we are OK with leaving things as is until someone provides actual fixes or whether we want to provide some aid in the meantime that can be easily reverted once it's fixed.
(This is similar to what I tell people who voice it would be better to more or less only focus on `rattler-build` and not not invest in `conda-build` anymore; until we have a solution ready for direct deployment, we need pragmatic approaches.)

Do note that the changes herein are intentionally structured such that they can easily be reverted:
1. Different changes are put into separate commits so can be reverted (and also cherry-picked if we want to split this),
2. Even if single commits can't be automatically reverted, one only has to rename function names/calls (`_select_lines` --> `select_lines`; `re_*` --> `re.*`) or re-indent code (in `Config.copy`) and remove the then extraneous code.
3. The actual workings are unchanged without any hard to follow restructuring (which is what makes point 2 possible).

These workarounds, of course, do unfortunately add maintenance burden.
In my opinion, due to the above points and the narrowed scope of the changes, it is an acceptable burden to take.
That is, not in the general sense, but only in light of the strain on the (conda-forge's) ecosystem this is trying to alleviate.

----
That being said, what this PR needs is definitely more explanatory comments (if we add workarounds, we should mark them as such and explain, e.g., why the heck we'd add 50 lines only to be able to memoize `select_lines`) and ideally someone reproduce/confirm the impact of these changes (since we have to gauge impact vs added maintenance).

**Comment by mbargull** [comment]
> [The deepcopy code](https://github.com/python/cpython/blob/main/Lib/copy.py#L135-L158) is a slow way to copy Python objects, json loads/dumps or pickle are known optimizations.

I opt'ed to not use `json` for this since, depending on your data, it is a destructive process.
E.g., not only would it not capture instances of custom classes but also `loads(dumps({1:2})) == {"1": 2} != {1: 2}`.

File: conda_build/metadata.py (hunk 1 above)
Comment (mbargull) [reply conda_build/metadata.py:117]: As mentioned by @dholth below, the `re` module already caches compiled patterns.
In contrast to `re`'s functions, the wrappers here not only cache for the pattern but also the matched string.
This is because we have redundant `re.*(same_pattern, same_string)` calls due to the repeated re-processing mentioned in https://github.com/conda/conda-build/issues/5224#issue-2179165566 .
As such, the memoizations here not try to sidestep the recompilation of patterns but the actual regex search/matches.

File: conda_build/metadata.py (hunk 4 above)
Comment (mbargull) [reply conda_build/metadata.py:135]: Personally, I'm also triggered by unnecessary repetitions, but I'm fine with it in this case since it should be a set-and-forget thing.

That said, I would make sense to profile this (for majorly affected cases) again to see which of these memoization wrappers have individually significant impact.
Meaning, we possibly only need a subset of them to get roughly the same results.

File: conda_build/metadata.py (hunk 5 above)
Comment (mbargull) [reply conda_build/metadata.py:328]: +1 Makes sense to add/extend tests for `select_lines` which ensure repeated calls to `select_lines` and the non-memoized `_select_lines` always yield the same results.

Comment (kenodegard) [reply conda_build/metadata.py:328]: Enabling CodSpeed benchmarks and adding a new `test_select_lines_battery` test in https://github.com/conda/conda-build/pull/5233 so we can better measure improvements

**Comment by kenodegard** [comment]
Thanks everyone for their input here. In an effort to get this discussion moving again (and to help us make data driven decisions with benchmarks) this PR is superseded by several other PRs.

Smaller PRs with isolated changes for easier review and clearer understanding of which are syntax improvements versus speed improvements:
- https://github.com/conda/conda-build/pull/5253
- https://github.com/conda/conda-build/pull/5278
- https://github.com/conda/conda-build/pull/5279
- https://github.com/conda/conda-build/pull/5280
- https://github.com/conda/conda-build/pull/5281
- https://github.com/conda/conda-build/pull/5283

And an alternative solution to the `select_lines` caching proposed here:
- https://github.com/conda/conda-build/pull/5237
